Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+///
+/// ### Example
+///
+/// ```swift
+/// // Swift source code with issues
+/// class Iterator: {
+///   var counter = 0
+///   var wasCounterCalled: Bool { counter > 0 }
+///
+///   init(counter: Int = 0) {
+///     self.counter = counter
+///   }
+///
+///   // Description of the `loopOver` function
+///   func loopOver() {
+///     for (i = 0; i != 10; i += 1) { }
+///   }
+/// }
+///
+/// // Diagnostics produced by a parser could be applied to produce an annotated output like:
+/// """
+///  1 │ class Iterator: {
+///    │                 ╰─ error: expected type in inherited type
+///  2 │   var counter = 0
+///  3 │   var wasCounterCalled: Bool { counter > 0 }
+///    ┆
+///  9 │   // Description of the `loopOver` function
+/// 10 │   func loopOver() {
+/// 11 │     for (i = 0; i != 10; i += 1) { }
+///    │         ~~~~~~~~~~~~~~~~~~~~~~~~
+///    │     │   │  ╰─ error: expected ')' to end tuple pattern
+///    │     │   ╰─ note: to match this opening '('
+///    │     ╰─ error: C-style for statement has been removed in Swift 3
+/// 12 │   }
+/// 13 │ }
+/// """
+/// ```
+///
+/// ### Motivation
+///
+/// This formatter enables better development practices by:
+///
+/// - **Improving Debugging**: Diagnostic messages, including errors, warnings, notes, fix-its, highlights, and remarks,
+/// are aligned in context to the source code, making it easier to understand what needs to be corrected or considered.
+///
+/// - **Enhanced Readability**: The formatter enriches the source code with line numbers,
+/// annotations, and optional text suffixes, providing a full picture to the developer.
+///
+/// - **Flexible Customization**: Through the `DiagnosticDecorator` protocol, you can customize
+/// the appearance of diagnostic output.
 public struct DiagnosticsFormatter {
-
-  /// A wrapper struct for a source line, its diagnostics, and any
-  /// non-diagnostic text that follows the line.
-  private struct AnnotatedSourceLine {
-    var diagnostics: [Diagnostic]
-    var sourceString: String
-
-    /// Non-diagnostic text that is appended after this source line.
-    ///
-    /// Suffix text can be used to provide more information following a source
-    /// line, such as to provide an inset source buffer for a macro expansion
-    /// that occurs on that line.
-    var suffixText: String
-
-    /// Whether this line is free of annotations.
-    var isFreeOfAnnotations: Bool {
-      return diagnostics.isEmpty && suffixText.isEmpty
-    }
+  /// An enumeration that describes the amount of contextual lines to display around each diagnostic message.
+  ///
+  /// - `limited(Int)`: Display a limited number of lines around each diagnostic. The integer value specifies the number of lines before and after the diagnostic message.
+  /// - `full`: Display the entire source code along with the diagnostics. Useful for getting a full overview, but could result in a large output.
+  enum ContextRange {
+    case limited(Int)
+    case full
```

Comment:
When do you think the full context range could be useful? I think recently, we have been aiming to only display a single error/warning + its inline buffer. If we display multiple errors inside the same context, it’s also not clear which notes belong to which errors.

CC @DougGregor

Reply:
My thinking behind offering the full context range was to provide end-developers with extensive customization options. For example, [the pointfree project uses `contextSize` set to max number of lines to achieve a full context view](https://github.com/pointfreeco/swift-macro-testing/blob/15916c0c328339f54c15d616465d79700e3f7de8/Sources/MacroTesting/AssertMacro.swift#L205).

Reply:
Ah, I see. I still have concerns about how this interacts with the notes but maybe it’s good enough even if the connection between notes and errors/warnings is not obvious.

Reply:
I'm not opposed to having the full context range. From a user perspective, I do wonder if I want to have a different context range for the top-level diagnostic vs. diagnostics in nested buffers (e.g., for macro expansions), because the nested buffers generally have content that's been synthesized.